rsconnect_jupyter/static/connect.js

```diff
@@ -147,6 +151,121 @@ define([
     $parent.find('.help-block').text('');
   }
 
+  /**
+   * FileListItemManager creates a manager for the file list dropdown.
+   * @param $listGroup {jQuery} the element of the added-files list
+   * @param $dropDown {jQuery} the element of the dropdown list
+   * @param availableFiles {Array<String>} the list of available files
+   * @param fileList {Array<String>} the list of files staged for deploy
+   * @param contents {Array<Object>} content list from the content manager
+   * @param basePath {String} base path of the notebook
+   * @param notebookPath {String} the notebooke path from `Jupyter.notebook.notebook_path`
+   * @returns {Object}
+   * @constructor
+   */
+  function FileListItemManager($listGroup, $dropDown, availableFiles, fileList, contents, basePath, notebookPath) {
+    return {
+      $listGroup: $listGroup,
+      $dropDown: $dropDown,
+      availableFiles: availableFiles,
+      fileList: fileList,
+      contents: contents,
+      basePath: basePath,
+      notebookPath: notebookPath,
+      /**
+       * pathSanitizer removes the basePath from the given path.
+       * @param path {string} path to sanitize
+       * @returns {string} path with basepath removed
+       */
+      pathSanitizer: function(path) {
+        // Assumption in here is that `path` contains `this.basePath`
+        return path.slice(this.basePath.length+1);
+      },
+      /**
+       * addAllFiles adds all files from the dropdown to the file list
+       */
+      addAllFiles: function() {
+        this.fileList = this.fileList.concat(this.availableFiles);
+        this.availableFiles = [];
+        this.updateDropDownItems();
+        this.updateListGroupItems();
+      },
+      /**
+       * updateListGroupItems updates the list group from the list of staged files
+       * @private Do not use outside of the FileListManager. If you need to do
+       * something specific that has to update the list group items, implement
+       * it as part of the manager. If you don't have access to reference vars
+       * that you need for it, add them to the constructor.
+       */
+      updateListGroupItems: function() {
+        var that = this;
+        that.$listGroup.empty();
+        that.contents.forEach(function (item) {
+          if (that.fileList.indexOf(item.path) !== -1) {
+            var li = document.createElement('li');
+            var i = document.createElement('i');
+            i.className = 'fa fa-times';
+            li.className = 'list-group-item';
+            li.appendChild(i);
```

Comment:
This is fine as is, I'm just curious why you didn't create the elements with jQuery?

Reply:
It's a habit I've gotten into that, in some instances, helps to minimize browser HTML parsing passes. It's kind of an overoptimization unless you do it by habit.